### PR TITLE
Support for 128 bit constants in dpdk backend

### PR DESCRIPTION
**Pull request by Sosutha**
Signed-off-by : Sosutha Sethuramapandian <sosutha.sethuramapandian@intel.com>

* Added support for 128 bit constants usage in dpdk backend
* Updated testcases and outputs

### CODE REVIEW

**Comment by fruffy** [comment]
Great! @jafingerhut Do we have manual STF or PTF tests for the DASH program that could check for correctness here? We tried to automate this process once but removed the automated DPDK PTF tests because of stability issues.

**Comment by jafingerhut** [comment]
> Great! @jafingerhut Do we have manual STF or PTF tests for the DASH program that could check for correctness here? We tried to automate this process once but removed the automated DPDK PTF tests because of stability issues.

There is an IPDK environment that you can run the DPDK software switch inside of, which adds a P4Runtime API server, but as far as I know the open source version of that has not been updated since July 2023.  I could be wrong about that.

Without using that P4Runtime API server, my belief is that the there is only a native P4 DPDK control plane API that one would need to use in order to add table entries.  I have not used that API before, nor even where to find documentation for it, but I haven't looked hard.

File: backends/dpdk/dpdkHelpers.cpp
Comment (syogender) [inline backends/dpdk/dpdkHelpers.cpp:635]: What happen for value greater than 64 bits and less than 128 bits

File: backends/dpdk/dpdkArch.cpp
Comment (syogender) [inline backends/dpdk/dpdkArch.cpp:2965]: how constant not fitting in uint64 will be handled?

Comment (Sosutha) [reply backends/dpdk/dpdkArch.cpp:2965]: [backends/dpdk/dpdkHelpers.cpp] - This file has the changes in this same PR where Unsigned values more than 64 bit is handled.

**Review by syogender — APPROVED** [review]
lgtm

**Comment by Sosutha** [comment]
@fruffy @jafingerhut Could you please approve this PR for merging

**Comment by KrisNey-MSFT** [comment]
Will we try executing the spec file to check whether this is working?  bmv2 and dpdk targets/backends.  @chrispsommers for visibility
1. Are DASH blockers fixed, and 
2. can we run P4 DPDK and execute the DASH code in the repo?  

We can pick this up in the Behavioral Model call tomorrow Nov 14 2024

File: backends/dpdk/dpdkHelpers.cpp
Comment (fruffy) [reply backends/dpdk/dpdkHelpers.cpp:635]: Are you throwing an error for these cases? If so, where?

File: backends/dpdk/dpdkArch.cpp
Comment (fruffy) [reply backends/dpdk/dpdkArch.cpp:2965]: Why can the 128 bit case not be handled here? I am a little concerned that we might be masking failures here.

Also: the error message at line 2977 is incorrect and probably similar error messages, too. It looks like the DPDK backend only supports bit widths with power of 2, is that correct?

File: backends/dpdk/dpdkArch.cpp
Comment (fruffy) [inline backends/dpdk/dpdkArch.cpp:2963]: 128 should be static constant that you define, maybe in `dpdkHelpers.h`. It should be used everywhere you use "128".

File: backends/dpdk/dpdkHelpers.cpp
Comment (Sosutha) [reply backends/dpdk/dpdkHelpers.cpp:635]: It will be thrown in dpdkArch.h itself, These cases will not reach this point of code.

File: backends/dpdk/dpdkArch.cpp
Comment (Sosutha) [reply backends/dpdk/dpdkArch.cpp:2965]: Code generation for 128 bit constant is handled in dpdkHelpers.cpp since here only the constraints needed for the DPDK is checked in this file.  DPDK backend supports till 64 bits and assembly instructions to handle only 128 bit after 64 bits. Hence the intermediate number of bits are ignored

File: backends/dpdk/dpdkArch.cpp
Comment (Sosutha) [reply backends/dpdk/dpdkArch.cpp]: Updated

Comment (fruffy) [reply backends/dpdk/dpdkArch.cpp]: I would consider wrapping this in a namespace or struct to avoid conflicts. You can then access the value like `DpdkConstants::SupportedInstructionWidth` or something similar. Capturing what this restriction is meant to represent.

**Review by fruffy — APPROVED** [review]
Approving to unblock things but would be good if @jafingerhut could give this another pass before merging. It's hard for me to tell whether this is  doing the right thing or implements the intended outcome.

**Comment by jafingerhut** [comment]
So I tried the attached P4 program, which requires you to use `-DBITWIDTH=<number>` on the command line to compile it, with all bit widths from 1 to 128, inclusive.

For bit widths 1 through 64, 72, 80, 88, 96, 104, 112, 120, and 128, it gave no error, and seemed to produce a reasonable .spec file, at least as long as assigning an action parameter value with width > 64 bits but < 128 bits to a 128-bit field is supposed to be legal in a spec file.

However, for other bit widths, i.e. those greater than 64 bits but less than 128 bits that were not a multiple of 8, it gave a CompilerBug error like the example below:
```
$ $HOME/forks2/p4c/build/p4c-dpdk -DBITWIDTH=65 --target dpdk --arch pna -o assign-bitX-to-bit128.spec assign-bitX-to-bit128.p4
terminate called after throwing an instance of 'P4::Util::CompilerBug'
  what():  In file: /home/andy/forks2/p4c/backends/dpdk/dpdkUtils.cpp:157
Compiler Bug: Metadata bit-width expected to be within 64-bits, found 65

Aborted (core dumped)
```
First, it should not core dump or get to a Compiler Bug.

Second, if you want it to give an error message, the message seems misleading, because it seems to be supporting widths larger than 64 bits, just not all bit widths in the range 65 to 128.

Third, if it actually works for bit widths 72, 80, 88, 96, 104, 112, 120, why not all of the others in the range 65 to 127?

**Comment by KrisNey-MSFT** [comment]
@maheswari-s - should I close this one as it has been overwritten?  Pls let us know :)

**Comment by jafingerhut** [comment]
Closing this PR, as #5074 is preferred over this one.

**Pull request closed without merging**
